R/req-perform-stream.R

```diff
@@ -52,6 +52,49 @@ req_perform_stream <- function(req, callback, timeout_sec = Inf, buffer_kb = 64)
   )
 }
 
+#' @param n number of lines to read from the stream
+#' @rdname req_perform_stream
+#' @export
+req_perform_stream_lines <- function(req, callback, timeout_sec = Inf, n = 1L) {
+  check_request(req)
+
+  handle <- req_handle(req)
+  callback <- as_function(callback)
+
+  stopifnot(is.numeric(timeout_sec), timeout_sec > 0)
+  stop_time <- Sys.time() + timeout_sec
+
+  stream <- curl::curl(req$url, handle = handle)
+  open(stream, "rbf")
+  withr::defer(close(stream))
+
+  continue <- TRUE
+  buffer <- c()
+  while(continue && isIncomplete(stream) && Sys.time() < stop_time) {
+    buffer <- c(buffer, readBin(stream, raw(), 128))
+
+    new_lines <- which(buffer == charToRaw("\n"))
+    if (length(new_lines) >= n) {
+      lines <- rawToChar(head(buffer, new_lines[n]))
+      lines <- strsplit(lines, "\n")[[1L]]
+      buffer <- tail(buffer, -new_lines[n])
+      continue <- isTRUE(callback(lines))
+    }
+  }
+
+  lines <- strsplit(rawToChar(buffer), "\n")[[1L]]
+  callback(lines)
```

Comment:
Right. This needs some more logic around the `isIncomplete()` because we do need to call the callback in the situation where the stream is complete are there are leftover bytes in the `buffer`.